Реализовано каскадное удаление, в сервисах попробовал исправить метод…

Alexandr-01 opened the pull request:
…ы получения по id, а также методы update. в сервисе курорта при получении доступных курортов добавил проверку на дату (подходят только те курорты, которые еще не начались)

Alexandr-01 committed:
Реализовано каскадное удаление, в сервисах попробовал исправить методы получения по id, а также методы update. в сервисе курорта при получении доступных курортов добавил проверку на дату (подходят только те курорты, которые еще не начались)

kartashovaa commented on src/main/java/com/kyd3snik/travel/services/CityService.java:
```diff
@@ -54,14 +60,9 @@ public void update(City city) {
-        List<Resort> resortsDeparture = resortService.findByDepartureCity(city);
-        List<Hotel> hotels = hotelService.findByCity(city);
-        throwIfCantDelete(user, city, resortsArrival, resortsDeparture, userService.getAll());
+        throwIfCantDelete(user, getById(id), resortService.findByArrivalCity(city),
```
Нинада так, верни как было, было читабельнее
Еще у тебя getById 2 раза вызывается

Alexandr-01 replied:
Хорошо, я изначально думал, что это плохая идея, но что-то нехорошее меня потянуло за руку))

kartashovaa commented on src/main/java/com/kyd3snik/travel/services/CountryService.java:
```diff
@@ -39,12 +41,15 @@ public void addCountry(Country country) {
     public Country getById(Long id) {
-        return countryRepository.findById(id).get();
+        Optional<Country> country = countryRepository.findById(id);
+        if (country.isPresent())
```
там есть метод orElseThrow, и это все сокращается до 
return countryRepository.findById(id).orElseThrow(), но лучше
return countryRepository.findById(id).orElseThrow(() -> new IllegalStateException("Тут текст ошибки"))

kartashovaa commented on src/main/java/com/kyd3snik/travel/services/HotelService.java:
```diff
@@ -55,11 +59,8 @@ public void addHotelRoomToHotel(Hotel hotel, HotelRoom hotelRoom) {
-        Hotel hotel = getById(id);
-        List<Resort> resorts = resortService.findByHotel(hotel);
-        throwIfCantDelete(user, resorts);
+        throwIfCantDelete(user, resortService.findByHotel(getById(id)));
```
Не надо так сокращать, код менее читабельный

Alexandr-01 committed:
Поправил поиск по id и читабельность методов удаления

kartashovaa commented on src/main/java/com/kyd3snik/travel/services/CountryService.java:
```diff
@@ -56,19 +55,15 @@ public void delete(long id) {
-        throwIfCantDelete(user, users, cities, resortsArrival, resortsDeparture);
 
-        cityService.delete(cities);
+        throwIfCantDelete(user, userService.getAll(), cities, resortsArrival, resortsDeparture);
```
По идее, этот метод теперь бесполезный, так как связи настроены
Можешь это потестить, попробуй удалить страну, к которой привязан какой-нибудь параметр метода, например пользователь, если  все ок, то база должна кинуть исключение

Alexandr-01 replied:
мне кажется, что стоит оставить этот метод, т.к. он более четко обозначает ошибку

kartashovaa replied:
Ну её знаю, каким образом? Тем что ошибку кидает не база, а ты? Тем более он максимально медленно работает. Ты тут тянешь всех пользователей, города, для каждого города курорты, это очень много данных.

kartashovaa commented on src/main/java/com/kyd3snik/travel/services/ResortService.java:
```diff
@@ -103,8 +103,9 @@ public void delete(List<Resort> resorts) {
-    private List<Resort> filterPurchasedResorts(List<Resort> resorts) {
+    private List<Resort> filterPurchasedAndPastResorts(List<Resort> resorts) {
         return resorts.stream()
+                .filter(resort -> !resort.getStartDate().before(DateUtil.getToday()))
```
Условия с отрицанием читаются сложнее, в данном случае можно заменить на DateUtil.getToday().after(resort.getStartDate())

kartashovaa commented on src/main/java/com/kyd3snik/travel/controller/HotelController.java:
```diff
@@ -98,7 +99,7 @@ public String addHotel(
             @RequestParam HashMap<String, String> params) {
         City city = cityService.getById(idCity);
-        hotelService.addHotel(new Hotel(0, title, city, address, stars, new ArrayList<HotelRoom>()));
+        hotelService.addHotel(new Hotel(0, title, city, address, stars, new ArrayList<HotelRoom>(), Collections.emptyList()));
```
Раз уж тебе все равно править ПР, замени тут `new ArrayList<HotelRoom>()` на `Collections.emptyList()` для единообразия)

Alexandr-01 committed:
Изменил удаление и заменил методы сервиса на методы репозитория
убрал методы throwIfCantDelete в удалении и изменил методы в сервисе курорта (заменил приватные методы на методы репозитория)

kartashovaa committed:
Merge branch 'master' into cascade_deletion
